apps/emqx/src/emqx_persistent_session_ds/emqx_persistent_session_ds_router.erl

```diff
@@ -17,38 +17,48 @@
 -module(emqx_persistent_session_ds_router).
 
 -include("emqx.hrl").
+-include("emqx_mqtt.hrl").
 -include("emqx_ps_ds_int.hrl").
 
 -export([init_tables/0]).
 
 %% Route APIs
 -export([
-    do_add_route/2,
-    do_delete_route/2,
+    add_route/2,
+    add_route/3,
```

Comment:
I would question if exposing `add_route/3` is safe. Clients on different nodes (or in different zones) can choose to set it differently, leading to the following situation:  

- first client can be in a zone with upgrade_qos=true
- second one can be in a zone with upgrade_qos=false

Second subscriber doesn't expect QoS0 messages in the durable storage for the topic, while the first one can subscribe to the same topic with QoS1, which will lead to message duplication for the first client. I don't see message deduplication in the session code, correct me if I missed it.

But regardless, having QoS0 messages in both channels negates the advantage of this bypass, as the client still have to read and drop a bunch of messages.

Therefore, as I suggested here https://emqxio.slack.com/archives/C04H4RHKKEY/p1725653029483629 , upgrade_qos for durable sessions should be a _cluster-global_ setting. And as such, `add_route/2` function should decide the "scope" depending on this config and it should ignore client's desire.

Reply:
Another question: how does persistent_session router handle change of this setting?

Reply:
There's this [measure](https://github.com/emqx/emqx/pull/13819/files#diff-eb9115c096b866c464ff4ad37b615f6f62f9f4dd18980b974679ae27ee1a7400R338-R346). Currently, if any zone's or global MQTT config has `upgrade_qos = true`*, this logic of selective routing is outright disabled. I opted for such an approach, because 1) to have less discrepancies between session implementations, and 2) this option is used very rarely across our customers.

> And as such, `add_route/2` function should decide the "scope" depending on this config and it should ignore client's desire.

I don't think it's a good idea to have router concern itself with MQTT settings, it's session's domain after all. That's why it's currently exposed as an API detail.

---

* Granted, it doesn't currently handle situations where there's no `upgrade_qos = true` configured anywhere but some existing session has `upgrade_qos = true` in the subscription state, but I haven't found a sane way to handle this.

Reply:
> Another question: how does persistent_session router handle change of this setting?

Routes with different scopes are basically separate routes, there's no updates in that sense.

Reply:
> There's this [measure](https://github.com/emqx/emqx/pull/13819/files#diff-eb9115c096b866c464ff4ad37b615f6f62f9f4dd18980b974679ae27ee1a7400R338-R346).

1. Can a client call this API with wrong parameters due to bug, race condition or being restored from a very old state?
2. If not, how likely is such bug (or race condition) in the future?
3. If not, what if the session code is modified by someone else?
4. If such situation happen, does _one_ misbehaving client affect _all other_ clients subscribed to the same topic?

Reply:
P.S. QoS0 handling for shared subscriptions provides a legit case for the /3 API, but shared QoS0 subscription is very finicky question problem in itself. It requires study of MQTT spec.

Reply:
> If any answer is yes then this is not a very robust API.

Then there's little to no robust APIs. I'd also like to point that:
1. Currently, `upgrade_qos` is part of the session state.
2. Currently, `upgrade_qos` changes take effect at specific points in time, and that's also part of the session logic.

At the moment I don't see how all this can become part of the router. However, I'll try to fit everything better together so that this API becomes simpler.

Reply:
> Currently, upgrade_qos is part of the session state.

Previously, router wasn't QoS-selective. Now it is.